### PR TITLE
[Validator] Document constraints as php 8 Attributes

### PR DESCRIPTION
**Pull request by wkania**
We already have [examples](https://github.com/symfony/symfony-docs/pull/14230/files) of code for the PHP attributes.
We also have [Constraints as php 8 Attributes](https://github.com/symfony/symfony/pull/38309) and [2](https://github.com/symfony/symfony-docs/issues/14305).
The rest will be implemented in the [future](https://github.com/symfony/symfony/issues/38503).

### CODE REVIEW

**Comment by wkania** [comment]
@javiereguiluz Hi. Before I will add info about the attributes in constraints, could I receive some review about the message and position of the message?
In my opinion, it should be between the table and `Basic Usage` header.
The message:

    The ability to use PHP attributes to configure constraints was introduced in
    Symfony 5.2. Prior to this, Doctrine Annotations were the only way to
    annotate field of a class with constraint configuration.

**Comment by OskarStark** [comment]
I would love to get your review here @derrabus 👍  Thanks

@wkania I am not sure if we need this information, the code examples are only available in the `5.2` docs 🤔

File: reference/constraints/Callback.rst [hunk 1]
```diff
@@ -178,6 +195,18 @@ You can then use the following configuration to invoke this validator:
...
         }
 
+    .. code-block:: php-attributes
+
+        // src/Entity/Author.php
+        namespace App\Entity;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        #[Assert\Callback(["Acme\Validator", "validate"])]
```
Comment (derrabus) [inline reference/constraints/Callback.rst]:
```suggestion
        use Acme\Validator;
        use Symfony\Component\Validator\Constraints as Assert;

        #[Assert\Callback([Validator::class, 'validate'])]
```

* We should prefer `::class` over FQCN strings in PHP code.
* We should use single quotes if we don't need the additional semantics of double quotes.

File: reference/constraints/CardScheme.rst [hunk 2]
```diff
@@ -41,6 +41,22 @@ on an object that will contain a credit card number.
...
+        // src/Entity/Transaction.php
+        namespace App\Entity;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Transaction
+        {
+            #[Assert\CardScheme(
+                schemes: ["VISA"],
+                message: "Your credit card number is invalid."
```
Comment (derrabus) [inline reference/constraints/CardScheme.rst]:
```suggestion
                schemes: [Assert\CardScheme::VISA],
                message: 'Your credit card number is invalid.',
```

* We have added constants for the available card schemes and I suggest we make use of them.
* CS: prefer single quotes.
* CS: use trailing comma in multiline constructor calls.

File: reference/constraints/Choice.rst [hunk 3]
```diff
@@ -58,6 +58,27 @@ If your valid choice list is simple, you can pass them in directly via the
...
+        {
+            const GENRES = ['fiction', 'non-fiction'];
+
+            #[Assert\Choice(["New York", "Berlin", "Tokyo"])]
+            protected $city;
+
+            /**
+             * You can also directly provide an array constant to the "choices" option in the annotation
+             */
+            #[Assert\Choice(choices: Conference::GENRES, message: "Choose a valid genre.")]
```
Comment (derrabus) [inline reference/constraints/Choice.rst]:
```suggestion
            // You can also directly provide an array constant to the "choices" option in the annotation.
            #[Assert\Choice(choices: Author::GENRES, message: 'Choose a valid genre.')]
```

The code comment explains the following attribute block. I find it confusing to use a phpdoc block for that.

File: reference/constraints/Choice.rst [hunk 4]
```diff
@@ -160,6 +181,19 @@ constraint.
...
+    .. code-block:: php-attributes
+
+        // src/Entity/Author.php
+        namespace App\Entity;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Author
+        {
+            #[Assert\Choice(callback: "getGenres")]
```
Comment (derrabus) [inline reference/constraints/Choice.rst]:
```suggestion
            #[Assert\Choice(callback: 'getGenres')]
```

File: reference/constraints/Choice.rst [hunk 5]
```diff
@@ -225,6 +259,19 @@ you can pass the class name and the method as an array.
...
+    .. code-block:: php-attributes
+
+        // src/Entity/Author.php
+        namespace App\Entity;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Author
+        {
+            #[Assert\Choice(callback: ["App\Entity\Genre", "getGenres"])]
```
Comment (derrabus) [inline reference/constraints/Choice.rst]:
```suggestion
            #[Assert\Choice(callback: [Genre::class, 'getGenres'])]
```

File: reference/constraints/Count.rst [hunk 6]
```diff
@@ -47,6 +47,24 @@ you might add the following:
...
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Participant
+        {
+            #[Assert\Count(
+                min: 1,
+                max: 5,
+                minMessage: "You must specify at least one email",
+                maxMessage: "You cannot specify more than {{ limit }} emails"
```
Comment (derrabus) [inline reference/constraints/Count.rst]:
```suggestion
                minMessage: 'You must specify at least one email',
                maxMessage: 'You cannot specify more than {{ limit }} emails',
```

File: reference/constraints/Isbn.rst [hunk 7]
```diff
@@ -43,6 +43,22 @@ on an object that will contain an ISBN.
...
+        // src/Entity/Book.php
+        namespace App\Entity;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Book
+        {
+            #[Assert\Isbn(
+                type: "isbn10",
+                message : "This value is not valid."
```
Comment (derrabus) [inline reference/constraints/Isbn.rst]:
```suggestion
                type: Assert\Isbn::ISBN_10,
                message : 'This value is not valid.',
```

**Review by derrabus — CHANGES_REQUESTED** [review]
Thank you very much for working on the documentation of my feature.

The added code blocks look pretty much correct. However, I'd like to suggest a couple of code-style changes:

* Some constraints have worked with magic strings in the past (`CardScheme`, `Isbn`). While implementing the attributes I have introduced constants for those strings and my suggestion would be to use them.
* We should use `::class` instead of FQCN strings imho, but I don't know if there's a policy for the docs regarding that preference.
* Symfony's code-style prefers single quotes over double quotes. I'd suggest to apply this preference to attributes as well. They're PHP code, after all.
* I'd prefer to use trailing commas in multiline attributes. This would be consistent with Symfony's code style for multiline arrays and multiline Doctrine-style annotations.

File: reference/constraints/Email.rst [hunk 8]
```diff
@@ -37,6 +37,21 @@ Basic Usage
...
+
+        // src/Entity/Author.php
+        namespace App\Entity;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Author
+        {
+            #[Assert\Email(
+                message: "The email '{{ value }}' is not a valid email."
```
Comment (OskarStark) [inline reference/constraints/Email.rst:50]:
```suggestion
                message: "The email '{{ value }}' is not a valid email.",
```

File: reference/constraints/EqualTo.rst [hunk 9]
```diff
@@ -52,6 +52,24 @@ and that the ``age`` is ``20``, you could do the following:
...
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        class Person
+        {
+            #[Assert\EqualTo("Mary")]
+            protected $firstName;
+
+            #[Assert\EqualTo(
+                value: 20
```
Comment (OskarStark) [inline reference/constraints/EqualTo.rst:68]:
```suggestion
                value: 20,
```

File: reference/constraints/Expression.rst [hunk 10]
```diff
@@ -78,6 +78,22 @@ One way to accomplish this is with the Expression constraint:
...
+    .. code-block:: php-attributes
+
+        // src/Model/BlogPost.php
+        namespace App\Model;
+
+        use Symfony\Component\Validator\Constraints as Assert;
+
+        #[Assert\Expression(
+            "this.getCategory() in ['php', 'symfony'] or !this.isTechnicalPost()",
+            message: "If this is a tech post, the category should be either php or symfony!"
```
Comment (OskarStark) [inline reference/constraints/Expression.rst:90]:
```suggestion
            message: "If this is a tech post, the category should be either php or symfony!",
```

File: reference/constraints/Expression.rst [hunk 11]
```diff
@@ -163,6 +179,26 @@ more about the expression language syntax, see
...
+
+            use Symfony\Component\Validator\Constraints as Assert;
+
+            class BlogPost
+            {
+                // ...
+
+                #[Assert\Expression(
+                    "this.getCategory() in ['php', 'symfony'] or value == false",
+                    message: "If this is a tech post, the category should be either php or symfony!"
```
Comment (OskarStark) [inline reference/constraints/Expression.rst:195]:
```suggestion
                    message: "If this is a tech post, the category should be either php or symfony!",
```

**Review by derrabus — APPROVED** [review]

**Review by wkania — COMMENTED** [review]
@derrabus All fixed

> I would love to get your review here @derrabus +1 Thanks
> 
> @wkania I am not sure if we need this information, the code examples are only available in the `5.2` docs thinking

I do not know what is the policy behind adding such information, but @derrabus [added](https://github.com/symfony/symfony-docs/pull/14230/files) it.

File: reference/constraints/Choice.rst (hunk 3 above)
Comment (wkania) [reply reference/constraints/Choice.rst]: Just copy, paste from annotation example. Fixed

File: reference/constraints/Choice.rst (hunk 4 above)
Comment (wkania) [reply reference/constraints/Choice.rst]: Fixed

File: reference/constraints/Choice.rst (hunk 5 above)
Comment (wkania) [reply reference/constraints/Choice.rst]: Fixed

File: reference/constraints/Count.rst (hunk 6 above)
Comment (wkania) [reply reference/constraints/Count.rst]: Fixed

File: reference/constraints/Isbn.rst (hunk 7 above)
Comment (wkania) [reply reference/constraints/Isbn.rst]: Fixed

File: reference/constraints/Callback.rst (hunk 1 above)
Comment (wkania) [reply reference/constraints/Callback.rst]: Ok, good to know. I have created attributes examples based on the annotation. I will fix it everywhere. In some other pull request, I will also update the annotations.

**Comment by derrabus** [comment]
I don't think that we want to have this `versionadded` block on each constraint page. On some overview article that describes the various ways a developer might configure constraints, this block is certainly helpful, though.

File: reference/constraints/CardScheme.rst (hunk 2 above)
Comment (wkania) [reply reference/constraints/CardScheme.rst]: I added commas in all multiline constructor calls.

**Review by maxhelias — APPROVED** [review]

**Review by xabbuh — APPROVED** [review]

**Review by derrabus — APPROVED** [review]

**Comment by wkania** [comment]
Hi @OskarStark , @derrabus , @xabbuh , @maxhelias 
From what I see, @HeahDude  is a [former member of docs team](https://github.com/symfony/symfony-docs/pull/14824).
What about his review of this pull request?

**Review by OskarStark — APPROVED** [review]

**Comment by OskarStark** [comment]
It took time, but here we go, this is in now. Thank you very much @wkania.

**Pull request merged**